File: android/src/main/java/app/capgo/plugin/health/background/BackgroundSyncConfig.kt
package app.capgo.plugin.health.background

import app.capgo.plugin.health.HealthDataType
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import org.json.JSONArray
import org.json.JSONObject

data class BackgroundSyncApiRequestConfig(
    val url: String,
    val headers: Map<String, String> = emptyMap()
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("url", url)
        put("headers", JSONObject(headers))
    }

    fun toJSObject(): JSObject = JSObject().apply {
        put("url", url)
        put("headers", JSObject().apply {
            headers.forEach { (key, value) -> put(key, value) }
        })
    }

    companion object {
        fun fromJson(json: JSONObject): BackgroundSyncApiRequestConfig {
            val url = json.optString("url")
            require(url.isNotBlank()) { "Background sync URL is required." }

            val headersJson = json.optJSONObject("headers") ?: JSONObject()
            val headers = mutableMapOf<String, String>()
            val keys = headersJson.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                headers[key] = headersJson.optString(key)
            }
            return BackgroundSyncApiRequestConfig(url = url, headers = headers)
        }
    }
}

enum class BackgroundSyncInterval(
    val identifier: String,
    val intervalMinutes: Long
) {
    FIFTEEN_MINUTES("15min", 15L),
    THIRTY_MINUTES("30min", 30L),
    ONE_HOUR("1hour", 60L),
    EIGHT_HOURS("8hours", 480L),
    TWENTY_FOUR_HOURS("24hours", 1440L);

    companion object {
        fun from(identifier: String?): BackgroundSyncInterval? {
            return entries.firstOrNull { it.identifier == identifier }
        }
    }
}

data class BackgroundSyncConfig(
    val subjectId: String,
    val getLastSync: BackgroundSyncApiRequestConfig,
    val postSamples: BackgroundSyncApiRequestConfig,
    val dataTypes: List<HealthDataType>,
    val interval: BackgroundSyncInterval,
    val enabled: Boolean = false
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("subjectId", subjectId)
        put("getLastSync", getLastSync.toJson())
        put("postSamples", postSamples.toJson())
        put("dataTypes", JSONArray().apply {
            dataTypes.forEach { put(it.identifier) }
        })
        put("interval", interval.identifier)
        put("enabled", enabled)
    }

    fun toOptionsJSObject(): JSObject = JSObject().apply {
        put("subjectId", subjectId)
        put("getLastSync", getLastSync.toJSObject())
        put("postSamples", postSamples.toJSObject())
        put("dataTypes", JSArray().apply {
            dataTypes.forEach { put(it.identifier) }
        })
        put("interval", interval.identifier)
    }

    fun withEnabled(enabled: Boolean): BackgroundSyncConfig = copy(enabled = enabled)

    companion object {
        fun fromJson(json: JSONObject): BackgroundSyncConfig {
            val dataTypesJson = json.optJSONArray("dataTypes") ?: JSONArray()
            val dataTypes = mutableListOf<HealthDataType>()
            for (index in 0 until dataTypesJson.length()) {
                val identifier = dataTypesJson.optString(index)
                val dataType = HealthDataType.from(identifier)
                    ?: throw IllegalArgumentException("Unsupported background sync data type: $identifier")
                dataTypes.add(dataType)
            }

            val interval = BackgroundSyncInterval.from(json.optString("interval"))
                ?: throw IllegalArgumentException("Background sync interval must be one of: 15min, 30min, 1hour, 8hours, 24hours.")
            require(dataTypes.isNotEmpty()) { "Background sync requires at least one data type." }

            return BackgroundSyncConfig(
                subjectId = json.optString("subjectId").also {
                    require(it.isNotBlank()) { "Background sync subjectId is required." }
                },
                getLastSync = BackgroundSyncApiRequestConfig.fromJson(
                    json.optJSONObject("getLastSync")
                        ?: throw IllegalArgumentException("Background sync getLastSync configuration is required.")
                ),
                postSamples = BackgroundSyncApiRequestConfig.fromJson(
                    json.optJSONObject("postSamples")
                        ?: throw IllegalArgumentException("Background sync postSamples configuration is required.")
                ),
                dataTypes = dataTypes.distinct(),
                interval = interval,
                enabled = json.optBoolean("enabled", false)
            )
        }
    }
}

File: ios/Sources/HealthPlugin/background/BackgroundSyncConfig.swift
import Foundation

struct BackgroundSyncApiRequestConfig: Codable, Equatable {
    let url: String
    let headers: [String: String]

    static func from(dictionary: [String: Any], key: String) throws -> BackgroundSyncApiRequestConfig {
        guard let raw = dictionary[key] as? [String: Any] else {
            throw BackgroundSyncConfigError.missingKey(key)
        }
        guard let url = raw["url"] as? String, !url.isEmpty else {
            throw BackgroundSyncConfigError.invalidValue("\(key).url")
        }
        var headers: [String: String] = [:]
        if let headersRaw = raw["headers"] as? [String: Any] {
            for (headerKey, value) in headersRaw {
                if let stringValue = value as? String {
                    headers[headerKey] = stringValue
                }
            }
        }
        return BackgroundSyncApiRequestConfig(url: url, headers: headers)
    }
}

enum BackgroundSyncInterval: String, Codable, CaseIterable {
    case fifteenMinutes = "15min"
    case thirtyMinutes = "30min"
    case oneHour = "1hour"
    case eightHours = "8hours"
    case twentyFourHours = "24hours"

    var intervalMinutes: Int {
        switch self {
        case .fifteenMinutes: return 15
        case .thirtyMinutes: return 30
        case .oneHour: return 60
        case .eightHours: return 480
        case .twentyFourHours: return 1440
        }
    }

    static func from(identifier: String?) -> BackgroundSyncInterval? {
        guard let identifier = identifier else { return nil }
        return BackgroundSyncInterval(rawValue: identifier)
    }
}

struct BackgroundSyncConfig: Codable, Equatable {
    let subjectId: String
    let getLastSync: BackgroundSyncApiRequestConfig
    let postSamples: BackgroundSyncApiRequestConfig
    let dataTypes: [String]
    let interval: BackgroundSyncInterval
    var enabled: Bool

    func withEnabled(_ enabled: Bool) -> BackgroundSyncConfig {
        var copy = self
        copy.enabled = enabled
        return copy
    }

    static func from(pluginCall dictionary: [String: Any]) throws -> BackgroundSyncConfig {
        guard let subjectId = dictionary["subjectId"] as? String, !subjectId.isEmpty else {
            throw BackgroundSyncConfigError.missingKey("subjectId")
        }
        let getLastSync = try BackgroundSyncApiRequestConfig.from(dictionary: dictionary, key: "getLastSync")
        let postSamples = try BackgroundSyncApiRequestConfig.from(dictionary: dictionary, key: "postSamples")
        guard let dataTypesRaw = dictionary["dataTypes"] as? [String], !dataTypesRaw.isEmpty else {
            throw BackgroundSyncConfigError.invalidValue("dataTypes")
        }
        let dataTypes = Array(Set(dataTypesRaw.filter { !$0.isEmpty }))
        guard !dataTypes.isEmpty else {
            throw BackgroundSyncConfigError.invalidValue("dataTypes")
        }
        for identifier in dataTypes {
            guard HealthDataType(rawValue: identifier) != nil else {
                throw BackgroundSyncConfigError.unsupportedDataType(identifier)
            }
        }
        guard let intervalRaw = dictionary["interval"] as? String,
              let interval = BackgroundSyncInterval.from(identifier: intervalRaw) else {
            throw BackgroundSyncConfigError.invalidValue("interval")
        }
        return BackgroundSyncConfig(
            subjectId: subjectId,
            getLastSync: getLastSync,
            postSamples: postSamples,
            dataTypes: dataTypes,
            interval: interval,
            enabled: false
        )
    }
}

enum BackgroundSyncConfigError: LocalizedError {
    case missingKey(String)
    case invalidValue(String)
    case unsupportedDataType(String)

    var errorDescription: String? {
        switch self {
        case let .missingKey(key):
            return "Background sync \(key) configuration is required."
        case let .invalidValue(key):
            return "Background sync \(key) is invalid."
        case let .unsupportedDataType(identifier):
            return "Unsupported background sync data type: \(identifier)"
        }
    }
}
